scripts/codegen/modules/devicedeclare.py

```diff
@@ -0,0 +1,391 @@
+                                                device_init,
+                                                device_level,
+                                                device_prio,
+                                                device_api,
+                                                device_info)
+        devices_declared.append(device_declared)
+
+    if True not in devices_declared:
+        err = "No active device found for {} = {} and {}.".format(
+            ', '.join(device_configs), ', '.join(devices_declared),
+            ', '.join(driver_names))
+        codegen.log(err)
+        raise codegen.Error(err)
+
+
+def _device_generate_struct(type_of_struct, _struct):
+    struct = ""
+
+    # convert _struct into a list. Struct might have only one element
+    if type(_struct) is str:
+        _struct = [_struct]
+    else:
+        _struct = list(_struct)
+
+    if type_of_struct == 'config':
+        struct += 'static const struct {} ${{device-config-info}} = {{\n'.format(_struct[0])
+    elif type_of_struct == 'data':
+        struct += 'static struct {} ${{device-data}} = {{\n'.format(_struct[0])
+    else:
+        msg("Not expected")
+
+    if len(_struct) > 1:
+        struct += _struct[1]
+
+    struct += '};\n\n'
+    return struct
+
+
+def _device_generate_irq_bootstrap(irq_names, irq_flag, irq_func):
+    irq_bootstrap_info = \
+        '#ifdef {}\n'.format(irq_flag) + \
+        'DEVICE_DECLARE(${device-name});\n' + \
+        'static void ${device-config-irq}(struct device *dev)\n' + \
+        '{\n'
+    for irq_name in irq_names:
+        irq_num = '${{interrupts/{}/irq}}'.format(irq_name)
+        irq_prio = '${{interrupts/{}/priority}}'.format(irq_name)
+        irq_bootstrap_info += \
+            '\tIRQ_CONNECT({},\n'.format(irq_num) + \
+            '\t\t{},\n'.format(irq_prio) + \
+            '\t\t{}_{},\n'.format(irq_func, irq_name) + \
+            '\t\tDEVICE_GET(${device-name}),\n' + \
+            '\t\t0);\n' + \
+            '\tirq_enable({});\n\n'.format(irq_num)
+    irq_bootstrap_info += \
+        '}\n' + \
+        '#endif /* {} */\n\n'.format(irq_flag)
+    return irq_bootstrap_info
+
+
+def device_declare(compatibles, init_prio_flag, kernel_level, irq_func,
+                   init_func, api, data_struct, config_struct):
+    _prio = init_prio_flag
+    _level = kernel_level
+    _isr = irq_func
+    _device_init = init_func
+    _api = api
+    _data_struct = data_struct
+    _config_struct = config_struct
+
+    device_configs = []
+    driver_names = []
+    device_info = ""
+    irq_names = None
+
+    for device_id in codegen.edts().device_ids_by_compatible(compatibles):
+        driver_name = codegen.edts().device_property(device_id, 'label')
+        device_configs.append("CONFIG_{}".format(driver_name.strip('"')))
+        driver_names.append(driver_name)
+        if irq_names is None:
+            # All devices should have the same irq names
```

Comment:
If you think the special case is necessary you can loop over all device_ids and create a device_info by instance. Feeding in the end to device_declare_single (instead of multi).

I don´t know and took what looked most like the way you have done it. Feel free to change.